test(protocol-designer): Cypress-Transfer-Platereader-Helper

alexjoel42 opened the pull request:
# Overview

1. Add plate reader smoke test
2. Add Transfer 1:1 test
3. Add helper functions to test if we like that pattern to reduce work used in createNewFlex.cy.ts
## Test Plan and Hands on Testing

## Changelog
Changed some of the selectors to work with PD 8.4.3 
<!--
List changes introduced by this PR considering future developers and the end user. Give careful thought and clear documentation to breaking changes.
-->

## Review requests
@y3rsh 

## Risk assessment
Going to likely need to look through selectors when we get to liquid class work because it will change fields. That being said, having automation that can be tweaked to liquid classes is helpful.

alexjoel42 committed:
fixed all selectors

koji commented:
@alexjoel42 
modified the title a little bit

y3rsh commented on protocol-designer/cypress/support/MixSteps.ts:
```diff
@@ -111,7 +111,7 @@ enum MixLocators {
   RenameBtn = 'button:contains("Rename")',
   StepNameInput = '[class="InputField__StyledInput-sc-1gyyvht-0 cLVzBl"]',
-  StepNotesInput = '[class^="RenameStepModal__DescriptionField"]',
+  StepNotesInput = '[class^="RenameStepModal__DescriptionField-sc-1k5vjxe-0 lkzOSf"]',
```
This needs to remain a class starts with, which is what `class^=` means and not have the generated part, or it needs to change altogether.

alexjoel42 replied:
I changed it to  
cy.contains('Step Notes')
        .next('textarea') // Find the textarea that is the next sibling
        .type('This is testing cypress automation in PD')

y3rsh commented on protocol-designer/cypress/support/MixSteps.ts:
```diff
@@ -72,7 +72,7 @@ enum MixContent {
   StepNotes = 'Step Notes',
   CypressTest = 'Cypress Mix Test',
-  TouchTipFromBottom = 'Touch tip position from bottom',
+  TouchTipFromTop = 'Touch tip position from bottom',
```
The variable and the text do not match.

y3rsh commented on protocol-designer/cypress/support/MixSteps.ts:
```diff
@@ -638,7 +638,7 @@ export const MixVerifications = {
       cy.log('Verifying TouchTip ---------------------------')
       cy.contains(MixContent.TouchTip).should('exist').should('be.visible')
-      cy.contains(MixContent.TouchTipFromBottom)
+      cy.contains(MixContent.TouchTipFromTop)
```
Something is up here, verify in the cypress UI that you are verifying in tough Tip when this code is running.

alexjoel42 replied:
We changed touch tip's starting location recently in the next which is why I had to change touch tip from bottom --> touch tip from top

y3rsh commented on protocol-designer/cypress/support/ModuleSteps.ts:
```diff
@@ -116,15 +113,53 @@ export const ModuleSteps = {
+      cy.contains('Absorbance Plate Reader').click()
+    },
+  }),
+  DefineInitilizationSingle: (): StepThunk => ({
+    call: () => {
+      // Goes through all the wavelengths
+      cy.contains('450 nm (blue)').click()
+      cy.contains('562 nm (green)').click()
+      cy.contains('562 nm (green)').click()
+      cy.contains('600 nm (orange)').click()
+      cy.contains('600 nm (orange)').click()
+      cy.contains('650 nm (red)').click()
+      cy.contains('650 nm (red)').click()
+      cy.contains('Other').click()
+    },
+  }),
```
The name of this step does not match what it does.

alexjoel42 replied:
Changed step name.

y3rsh commented on protocol-designer/cypress/support/ModuleSteps.ts:
```diff
@@ -116,15 +113,53 @@ export const ModuleSteps = {
+    call: () => {
+      // Goes through all the wavelengths
+      cy.contains('Custom wavelength')
+        .closest('div.Flex-sc-1qhp8l7-0.ivRgZg')
+        .find('input.InputField__StyledInput-sc-1gyyvht-0.cLVzBl')
```
Must improve these locators.

alexjoel42 replied:
This is hopefully a smidge better, but ideally I'd love to have test-ids for all of our test IDs 

```typescript 
cy.contains('Custom wavelength')
        .parents('.Flex-sc-1qhp8l7-0') // Go up to the parent container
        .find('input.InputField__StyledInput-sc-1gyyvht-0') // Find the input within the container
        .type('500') // Type the desired value 
```
Ideally we could do: 

```Typescript 
it('should type into the input field', () => {
    cy.get('[data-cy="wavelength-input"]')
      .type('500');

```

y3rsh commented on protocol-designer/cypress/support/ModuleSteps.ts:
```diff
@@ -133,9 +168,6 @@ export const ModuleVerifications = {
     },
   }),
 
-  /**
-   * Verify that the pause step has correct info in step preview (temp set to 4°C).
-   */
```
Add back the docstring, corrected

y3rsh commented on protocol-designer/cypress/support/SetupSteps.ts:
```diff
@@ -395,6 +409,24 @@ export const SetupSteps = {
     },
   }),
 
+  AddMoveStep: (): StepThunk => ({
+    call: () => {
+      cy.contains('button', 'Move').should('be.visible').click()
+    },
+  }),
+
+  UseGripperinMove: (): StepThunk => ({
+    call: () => {
+      cy.contains('button', 'Use gripper').should('be.visible').click()
+    },
+  }),
+
+  MoveToPlateReader: (): StepThunk => ({
+    call: () => {
+      cy.contains('button', 'Use gripper').should('be.visible').click()
+    },
+  }),
+
```
Add doc strings

y3rsh commented on protocol-designer/cypress/support/SetupSteps.ts:
```diff
@@ -530,6 +562,200 @@ export const SetupSteps = {
+  AddSourceLabwareDropdown: (): StepThunk => ({
+    call: () => {
+      cy.contains('Source labware')
+        .closest('div.Flex-sc-1qhp8l7-0.bsOFGI')
+        .find('div[tabindex="0"].Flex-sc-1qhp8l7-0.sc-bqWxrE')
```
These locators must be improved.

alexjoel42 replied:
cy.contains('Source labware')
        .parents()
        .contains('Choose option')
        .should('be.visible')
        .click()

y3rsh commented on protocol-designer/cypress/support/SetupSteps.ts:
```diff
@@ -730,3 +1016,85 @@ export const verifyCreateProtocolPage = (): void => {
+  if (heatershaker) {
+    SetupSteps.AddHeaterShaker().call()
+    SetupVerifications.HeaterShakerImg().call()
+  }
+
+  if (magblock) {
+    SetupSteps.AddMagBlock().call()
+    SetupVerifications.MagBlockImg().call()
+  }
+
+  if (tempdeck) {
+    SetupSteps.AddTempdeck2().call()
+    SetupVerifications.Tempdeck2Img().call()
+  }
+  if (platereader) {
+    SetupSteps.AddPlateReader().call()
+  }
+
+  SetupSteps.Confirm().call()
+  SetupSteps.Confirm().call()
+  SetupSteps.Confirm().call()
+  SetupSteps.EditProtocolA().call()
+}
+
+interface AddLabwareToDeckSlotCompoundInputs {
+  deckSlot?: string
+  labwareName?: string
+}
+export function AddLabwareToDeckSlotCompound(
+  string: AddLabwareToDeckSlotCompoundInputs
+): void {
+  const deckSlot = string.deckSlot ?? 'C3'
+  const labwareName = string.labwareName ?? 'Bio-Rad 96 Well Plate'
+
+  SetupSteps.ChoseDeckSlotWithLabware(deckSlot).call()
+  SetupSteps.AddHardwareLabware().call()
+  SetupSteps.ClickLabwareHeader().call()
+  SetupSteps.ClickWellPlatesSection().call()
+  SetupSteps.SelectLabwareByDisplayName(labwareName).call()
+}
```
See my PR into your branch 😺 https://github.com/Opentrons/opentrons/pull/17633

y3rsh reviewed (changes requested):
As you are getting this working, make sure to add cy.log in functions with even a little complexity as well as parameterized functions so troubleshooting is easier.

Also, do not add any more scope to this PR.  I would make small PRs that create one composite step at a time so that they may be easily reviewed. 

Just noticed another issue to fix in this PR when troubleshooting the below error.

- [x] We have 4 `describe` with the name `The Redesigned Create Protocol Landing Page` - rename them distincly
- [x]  We have 4 `it` with the same name `content and step 1 flow works`  - rename them distinctly

> Once my PR is merged and we have distinct names on each `describe` and `it`, we can see which test is causing this error!

```
[protocol-designer-tests]   The Redesigned Create Protocol Landing Page
[protocol-designer-tests]     1) content and step 1 flow works
[protocol-designer-tests]
[protocol-designer-tests]
[protocol-designer-tests]   0 passing (26s)
[protocol-designer-tests]   1 failing
[protocol-designer-tests]
[protocol-designer-tests]   1) The Redesigned Create Protocol Landing Page
[protocol-designer-tests]        content and step 1 flow works:
[protocol-designer-tests]      Error: The following error originated from your application code, not from Cypress.
[protocol-designer-tests]
[protocol-designer-tests]   > Uncaught Error: expected to find cutoutId from addressableAreaName undefined but could not
[protocol-designer-tests]
[protocol-designer-tests] When Cypress detects uncaught errors originating from your application it will automatically fail the current test.
[protocol-designer-tests]
[protocol-designer-tests] This behavior is configurable, and you can choose to turn this off by listening to the `uncaught:exception` event.
[protocol-designer-tests]
[protocol-designer-tests] https://on.cypress.io/uncaught-exception-from-application
[protocol-designer-tests]     at <unknown> (http://localhost:5178/src/utils/index.ts:118:5)
```

y3rsh committed:
fix(cypress): keep composite actions as step thunks (#17633)

@alexjoel42 Once you understand this, you can just merge it as it
targets your branch.

alexjoel42 committed:
fixed lead in and pulled in Y3rsh's composite step

alexjoel42 committed:
taking in notes

alexjoel42 committed:
Merge branch 'chore_release-pd-8.4.3' into Test(PD)-Cypress-Transfer-Platereader-Helper

jerader committed:
fix airGapInPlace bug with substepTimeline

jerader commented:
Leaving a comment here to explain the `substepTimeline` fix that was made:

With the new `airGapInPlace` command that was added recently, we only added support for it occurring on top of a waste chute/trash bin but forgot that it can also occur on top of a labware's well. The changes made were to account for it occurring over a labware well.

alexjoel42 committed:
We have the jerader fix in

alexjoel42 committed:
Merge branch 'chore_release-pd-8.4.3' into Test(PD)-Cypress-Transfer-Platereader-Helper

alexjoel42 committed:
Rename branch  (#17673)

Renaming branch to avoid PD build from failing.

alexjoel42 committed:
Cypres transfer platreader helper (#17676)

Trying to rename the branch
